### ISSUE TITLE
cloneRow exception

### ISSUE THREAD

**Issue by ddmo** [PHPOffice/PHPWord#950]
Hi, if $templateProcessor->cloneRow doesn't find the placeholder, go in exception.
How to prevent this behaviour? I don't know if in this particular template exist the placeHolder.

I try to catch the exception, but without results...

`
try {
  $template->cloneRow('placeHolder', "aaa");
} catch (Exception $e) {}
`

Thanks
Stefano

**Comment by jslirola** [comment]
Have you tried to print `$e->getMessage()`?

```
try {
	$templateProcessor->cloneRow('placeHolder', "aaa"); 
} catch (Exception $e) {
	echo $e->getMessage();
}
```

**Comment by ddmo** [comment]
Hi @jslirola, yes I've the same output, but this is the cloneRow output, not $e->getMessage() output. I need to continue the execution if template variable not found...

Thanks
Stefano

**Comment by samael205** [comment]
Hi, You need to continue because is a loop?

**Comment by ddmo** [comment]
Hi, no! Because if the placeholder is not present, the script call an exception... I need to continue.

Thanks
Stefano

**Comment by samael205** [comment]
Continue writing your code in the catch section
`try{
   	$template->cloneRow('placeHolder', "aaa"); 
} catch (Exception $e) {
	 // do nothing... php will ignore the exception and continue    
} `

**Comment by ddmo** [comment]
Hi Pedro, the code stop execution... I don't know the reason... 
This is the problem. 

Thanks
Stefano

**Comment by samael205** [comment]
Are you using TemplateProcessor class for the creation of the  object template?.

**Comment by ddmo** [comment]
Yes!



Il giorno 30 novembre 2016 @ 09:34:31, Pedro Jose (notifications@github.com)
ha scritto:

Are you using TemplateProcessor class for the creation of the object
template?.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/PHPOffice/PHPWord/issues/950#issuecomment-263813737>,
or mute the thread

**Comment by samael205** [comment]
What do you want to do?. I need see some code...

**Comment by ddmo** [comment]
Pedro, I create a method to create a doc from template with a lot of
placeholders (and some table row to clone). I can not know which
placeholders I will find in the template so, if a placeholder is not
present I need to continue the execution. The function
$templateProcessor->setValue(‘…’,’…') does
not generate an exception if the placeholder is not present. The
$templateProcessor->cloneRow(‘…’,’…’) generate an exception.

Thanks!
Stefano

Il giorno 30 novembre 2016 @ 09:39:45, Pedro Jose (notifications@github.com)
ha scritto:

What do you want to do?. I need see some code...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/PHPOffice/PHPWord/issues/950#issuecomment-263814735>,
or mute the thread

**Comment by samael205** [comment]
Which version is your Library of PhpWord?, The last?

**Comment by ddmo** [comment]
Yes, the develop branch.



Il giorno 30 novembre 2016 @ 10:15:48, Pedro Jose (notifications@github.com)
ha scritto:

Which version is your Library of PhpWord?, The last?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/PHPOffice/PHPWord/issues/950#issuecomment-263822123>,
or mute the thread

**Comment by samael205** [comment]
I've already seen your mistake. You are mistakenly changing the parameters
You are passing two strings to the method cloneRow $template->cloneRow('placeHolder', "aaa"); 
The cloneRow Method accepts two parameters: A string and a Integer.
The string is you placeHolder, that is right.
But the next parameter is the number of clones that you want.

**Comment by ddmo** [comment]
Ok Pedro, I’ve write a string instead of interior only for example. In my
code is correct.
Try with $template->cloneRow('placeHolder', 10) and you will get the same
behavior. If the placeholder is not present, che cloneRow method goes in
exception. The setValue method no.

thank you for your interest
Stefano



Il giorno 30 novembre 2016 @ 10:28:54, Pedro Jose (notifications@github.com)
ha scritto:

I've already seen your mistake. You are mistakenly changing the parameters
You are passing two strings to the method cloneRow
$template->cloneRow('placeHolder', "aaa");
The cloneRow Method accepts two parameters: A string and a Integer.
The string is you placeHolder, that is right.
But the next parameter is the number of clones that you want.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/PHPOffice/PHPWord/issues/950#issuecomment-263824933>,
or mute the thread

**Comment by samael205** [comment]
Because the setValue not thows any exception.

**Comment by ddmo** [comment]
Yes, I saw :) But, how to resolve this problem? I think the behavior of
setValue and cloreRow should be the same (without exception).

Thanks



Il giorno 30 novembre 2016 @ 10:47:20, Pedro Jose (notifications@github.com)
ha scritto:

Because the setValue not thows any exception.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/PHPOffice/PHPWord/issues/950#issuecomment-263829058>,
or mute the thread

**Comment by Rappi** [comment]
Hello,

I have this Problem too.
Is there a solution?

Rappi

**Comment by detrichter** [comment]
Hello, 
I'm using phpword version 0.17.0 and the behaviour of setValue and cloneRow is still as described in this thread. 

Using a try catch block did not work because the exception is already caught and thrown in the cloneRow method.

So my solution is to check weather the marker / tag / variable / placeholder exists in the template like so:
`$allVariables = $templateProcessor->getVariables();`
`if(in_array('placeHolder', $allVariables)){`
 `$templateProcessor->cloneRow('placeHolder', $numberOfClones);`
`}else{`
`// don't clone`
`}`

Perhaps it helps someone, even if the thread is old.
